File: dsl/packages/runtime/src/writes-effect.ts
/**
 * RFC0005: writes_effect expression language.
 *
 * This module provides types and an evaluator for expressing param-dependent
 * write effects in TaskSpec. The evaluator returns Exact(K) | May(K) | Unknown.
 */

// -----------------------------------------------------------------------------
// Effect Expression ADT
// -----------------------------------------------------------------------------

/** Keys{keyIds} -> always Exact({keys}) */
export interface EffectKeys {
  kind: "Keys";
  keyIds: number[];
}

/** FromParam(param) -> Exact if param constant, else Unknown */
export interface EffectFromParam {
  kind: "FromParam";
  param: string;
}

/** SwitchEnum(param, cases) -> Exact if param constant, May if bounded, else Unknown */
export interface EffectSwitchEnum {
  kind: "SwitchEnum";
  param: string;
  cases: Record<string, WritesEffectExpr>;
}

/** Union([e1, e2, ...]) -> combines effects */
export interface EffectUnion {
  kind: "Union";
  items: WritesEffectExpr[];
}

export type WritesEffectExpr =
  | EffectKeys
  | EffectFromParam
  | EffectSwitchEnum
  | EffectUnion;

// -----------------------------------------------------------------------------
// Evaluation Result
// -----------------------------------------------------------------------------

export interface ExactEffect {
  kind: "exact";
  keys: number[];
}

export interface MayEffect {
  kind: "may";
  keys: number[];
}

export interface UnknownEffect {
  kind: "unknown";
}

export type WritesEffect = ExactEffect | MayEffect | UnknownEffect;

// -----------------------------------------------------------------------------
// Gamma: compile/link-time environment
// -----------------------------------------------------------------------------

/** Maps param name to concrete value (number for key_id, string for enum) */
export type EffectGamma = Map<string, number | string>;

// -----------------------------------------------------------------------------
// Evaluator
// -----------------------------------------------------------------------------

/** Merge and sort key arrays, removing duplicates */
function mergeKeys(a: number[], b: number[]): number[] {
  const set = new Set([...a, ...b]);
  return [...set].sort((x, y) => x - y);
}

/** Combine two WritesEffect results */
function combineEffects(a: WritesEffect, b: WritesEffect): WritesEffect {
  // any Unknown => Unknown
  if (a.kind === "unknown" || b.kind === "unknown") {
    return { kind: "unknown" };
  }

  // Merge keys
  const merged = mergeKeys(a.keys, b.keys);

  // all Exact => Exact, else May
  if (a.kind === "exact" && b.kind === "exact") {
    return { kind: "exact", keys: merged };
  }

  return { kind: "may", keys: merged };
}

/**
 * Evaluate a writes_effect expression with given gamma context.
 * Returns Exact(keys), May(keys), or Unknown.
 */
export function evalWrites(
  expr: WritesEffectExpr,
  gamma: EffectGamma
): WritesEffect {
  switch (expr.kind) {
    case "Keys": {
      // Keys{} => always Exact with sorted, deduped keys (set semantics)
      const deduped = [...new Set(expr.keyIds)].sort((a, b) => a - b);
      return { kind: "exact", keys: deduped };
    }

    case "FromParam": {
      const val = gamma.get(expr.param);
      if (typeof val === "number") {
        // Param is known at compile/link time as a key_id
        return { kind: "exact", keys: [val] };
      }
      // Param not known => Unknown
      return { kind: "unknown" };
    }

    case "SwitchEnum": {
      const val = gamma.get(expr.param);
      if (typeof val === "string") {
        // Param is known as a string enum value
        const caseExpr = expr.cases[val];
        if (caseExpr) {
          return evalWrites(caseExpr, gamma);
        }
        // Known value but no matching case => Unknown
        return { kind: "unknown" };
      }

      // Param not constant => compute May(union of all cases)
      // If all cases are Exact or May, result is May(union), else Unknown
      const caseNames = Object.keys(expr.cases);
      if (caseNames.length === 0) {
        return { kind: "exact", keys: [] };
      }

      let allBounded = true;
      const allKeys = new Set<number>();

      for (const caseName of caseNames) {
        const caseExpr = expr.cases[caseName];
        const caseResult = evalWrites(caseExpr, gamma);
        if (caseResult.kind === "unknown") {
          allBounded = false;
          break;
        }
        for (const k of caseResult.keys) {
          allKeys.add(k);
        }
      }

      if (allBounded) {
        return { kind: "may", keys: [...allKeys].sort((a, b) => a - b) };
      }
      return { kind: "unknown" };
    }

    case "Union": {
      if (expr.items.length === 0) {
        return { kind: "exact", keys: [] };
      }

      let result: WritesEffect = { kind: "exact", keys: [] };
      for (const item of expr.items) {
        const itemResult = evalWrites(item, gamma);
        result = combineEffects(result, itemResult);
        if (result.kind === "unknown") {
          break; // Short-circuit on Unknown
        }
      }
      return result;
    }
  }
}

// -----------------------------------------------------------------------------
// Serialization (for parity testing)
// -----------------------------------------------------------------------------

/**
 * Serialize writes_effect to JSON object for parity testing.
 * Uses deterministic key ordering.
 */
export function serializeWritesEffect(
  expr: WritesEffectExpr
): Record<string, unknown> {
  switch (expr.kind) {
    case "Keys": {
      // Sort and dedupe for canonical output (set semantics)
      const deduped = [...new Set(expr.keyIds)].sort((a, b) => a - b);
      return { kind: "Keys", key_ids: deduped };
    }

    case "FromParam":
      return { kind: "FromParam", param: expr.param };

    case "SwitchEnum": {
      // Sort case keys for deterministic output
      const caseKeys = Object.keys(expr.cases).sort();
      const cases: Record<string, unknown> = {};
      for (const k of caseKeys) {
        cases[k] = serializeWritesEffect(expr.cases[k]);
      }
      return { kind: "SwitchEnum", param: expr.param, cases };
    }

    case "Union": {
      const items = expr.items.map((item) => serializeWritesEffect(item));
      return { kind: "Union", items };
    }
  }
}

// -----------------------------------------------------------------------------
// Builder helpers (for DSL use)
// -----------------------------------------------------------------------------

export function effectKeys(keyIds: number[]): EffectKeys {
  return { kind: "Keys", keyIds };
}

export function effectFromParam(param: string): EffectFromParam {
  return { kind: "FromParam", param };
}

export function effectSwitchEnum(
  param: string,
  cases: Record<string, WritesEffectExpr>
): EffectSwitchEnum {
  return { kind: "SwitchEnum", param, cases };
}

export function effectUnion(items: WritesEffectExpr[]): EffectUnion {
  return { kind: "Union", items };
}

File: engine/src/writes_effect.cpp
#include "writes_effect.h"
#include <algorithm>
#include <nlohmann/json.hpp>
#include <set>

namespace rankd {

// Helper for std::visit with multiple lambdas
template <class... Ts> struct overloaded : Ts... {
  using Ts::operator()...;
};
template <class... Ts> overloaded(Ts...) -> overloaded<Ts...>;

// Helper to merge and sort key vectors, removing duplicates
static std::vector<uint32_t> merge_keys(const std::vector<uint32_t> &a,
                                        const std::vector<uint32_t> &b) {
  std::set<uint32_t> merged(a.begin(), a.end());
  merged.insert(b.begin(), b.end());
  return std::vector<uint32_t>(merged.begin(), merged.end());
}

// Combine two WritesEffect results
static WritesEffect combine_effects(const WritesEffect &a,
                                    const WritesEffect &b) {
  // any Unknown => Unknown
  if (a.kind == EffectKind::Unknown || b.kind == EffectKind::Unknown) {
    return {EffectKind::Unknown, {}};
  }

  // all Exact => Exact, else May
  auto merged = merge_keys(a.keys, b.keys);

  if (a.kind == EffectKind::Exact && b.kind == EffectKind::Exact) {
    return {EffectKind::Exact, std::move(merged)};
  }

  return {EffectKind::May, std::move(merged)};
}

WritesEffect eval_writes(const WritesEffectExpr &expr,
                         const EffectGamma &gamma) {
  return std::visit(
      overloaded{
          [](const EffectKeys &e) -> WritesEffect {
            // Keys{} => always Exact with sorted, deduped keys (set semantics)
            std::set<uint32_t> deduped(e.key_ids.begin(), e.key_ids.end());
            return {EffectKind::Exact,
                    std::vector<uint32_t>(deduped.begin(), deduped.end())};
          },

          [&gamma](const EffectFromParam &e) -> WritesEffect {
            auto it = gamma.find(e.param);
            if (it != gamma.end() &&
                std::holds_alternative<uint32_t>(it->second)) {
              // Param is known at compile/link time as a key_id
              return {EffectKind::Exact, {std::get<uint32_t>(it->second)}};
            }
            // Param not known => Unknown
            return {EffectKind::Unknown, {}};
          },

          [&gamma](const EffectSwitchEnum &e) -> WritesEffect {
            auto it = gamma.find(e.param);
            if (it != gamma.end() &&
                std::holds_alternative<std::string>(it->second)) {
              // Param is known as a string enum value
              const auto &value = std::get<std::string>(it->second);
              auto case_it = e.cases.find(value);
              if (case_it != e.cases.end() && case_it->second) {
                return eval_writes(*case_it->second, gamma);
              }
              // Known value but no matching case => Unknown
              return {EffectKind::Unknown, {}};
            }

            // Param not constant => compute May(union of all cases)
            // If all cases are Exact, result is May(union), else Unknown
            if (e.cases.empty()) {
              return {EffectKind::Exact, {}};
            }

            bool all_bounded = true;
            std::set<uint32_t> all_keys;

            for (const auto &[case_name, case_expr] : e.cases) {
              if (!case_expr) {
                all_bounded = false;
                break;
              }
              WritesEffect case_result = eval_writes(*case_expr, gamma);
              if (case_result.kind == EffectKind::Unknown) {
                all_bounded = false;
                break;
              }
              all_keys.insert(case_result.keys.begin(), case_result.keys.end());
            }

            if (all_bounded) {
              return {EffectKind::May,
                      std::vector<uint32_t>(all_keys.begin(), all_keys.end())};
            }
            return {EffectKind::Unknown, {}};
          },

          [&gamma](const EffectUnion &e) -> WritesEffect {
            if (e.items.empty()) {
              return {EffectKind::Exact, {}};
            }

            WritesEffect result = {EffectKind::Exact, {}};
            for (const auto &item : e.items) {
              if (!item) {
                return {EffectKind::Unknown, {}};
              }
              WritesEffect item_result = eval_writes(*item, gamma);
              result = combine_effects(result, item_result);
              if (result.kind == EffectKind::Unknown) {
                break; // Short-circuit on Unknown
              }
            }
            return result;
          },
      },
      expr);
}

// Serialize writes_effect to JSON for manifest digest
std::string serialize_writes_effect(const WritesEffectExpr &expr) {
  nlohmann::ordered_json j;

  std::visit(
      overloaded{
          [&j](const EffectKeys &e) {
            j["kind"] = "Keys";
            // Sort and dedupe for canonical output (set semantics)
            std::set<uint32_t> deduped(e.key_ids.begin(), e.key_ids.end());
            j["key_ids"] =
                std::vector<uint32_t>(deduped.begin(), deduped.end());
          },

          [&j](const EffectFromParam &e) {
            j["kind"] = "FromParam";
            j["param"] = e.param;
          },

          [&j](const EffectSwitchEnum &e) {
            j["kind"] = "SwitchEnum";
            j["param"] = e.param;
            nlohmann::ordered_json cases_json;
            // Sort case keys for deterministic output
            std::vector<std::string> case_keys;
            for (const auto &[k, v] : e.cases) {
              case_keys.push_back(k);
            }
            std::sort(case_keys.begin(), case_keys.end());
            for (const auto &k : case_keys) {
              auto it = e.cases.find(k);
              if (it->second) {
                // Use ordered_json::parse to preserve key ordering
                cases_json[k] = nlohmann::ordered_json::parse(
                    serialize_writes_effect(*it->second));
              }
            }
            j["cases"] = cases_json;
          },

          [&j](const EffectUnion &e) {
            j["kind"] = "Union";
            nlohmann::ordered_json items_json = nlohmann::json::array();
            for (const auto &item : e.items) {
              if (item) {
                // Use ordered_json::parse to preserve key ordering
                items_json.push_back(nlohmann::ordered_json::parse(
                    serialize_writes_effect(*item)));
              }
            }
            j["items"] = items_json;
          },
      },
      expr);

  return j.dump();
}

} // namespace rankd
